ts/songParser.ts

```diff
@@ -0,0 +1,245 @@
+  /* Bb */
+  {"": "3,2,1,1", "m7sus4": "3,3,4,4", "madd9": "3,1,1,3", "dim7": "0,1,0,1", "aug7": "1,2,2,1", "aug": "3,2,2,1", "m7": "1,1,1,1", "m6": "0,1,1,1", "5": "3,x,1,1", "7": "1,2,1,1", "6": "0,2,1,1", "9": "3,2,4,3", "maj7": "3,2,1,0", "m9": "3,1,4,3", "sus2": "3,0,1,1", "sus4": "3,3,1,1", "2": "3,0,1,1", "11": "3,3,4,3", "dim": "3,1,0,1", "13": "1,2,3,1", "add4": "8,10,10,8", "add2": "3,2,1,3", "m7b5": "1,1,0,1", "msus4": "6,5,6,6", "m": "3,1,1,1", "add9": "3,2,1,3", "7sus4": "1,3,1,1"},
+  /* B  */
+  {"": "4,3,2,2", "m7sus4": "2,2,0,2", "madd9": "4,2,2,4", "dim7": "1,2,1,2", "aug7": "2,3,3,2", "aug": "4,3,3,2", "m7": "2,2,2,2", "m6": "1,2,2,2", "5": "x,x,2,2", "7": "2,3,2,2", "6": "1,3,2,2", "9": "4,3,5,4", "maj7": "4,3,2,1", "m9": "4,2,5,4", "sus2": "4,1,2,2", "sus4": "4,4,2,2", "2": "6,6,2,2", "11": "4,4,5,4", "dim": "4,2,1,2", "13": "2,3,4,2", "add4": "9,11,11,9", "add2": "6,6,7,6", "m7b5": "2,2,1,2", "msus4": "7,6,7,7", "m": "4,2,2,2", "add9": "4,3,2,4", "7sus4": "2,4,2,2"},
+];
+
+export const instrumentsData: { [instrument: string]: InstrumentData } = {
+  ukulele: {
+    strings: 4,
+    frets: 4,
+    chords: ukuleleChords
+  },
+  baritone: {
+    strings: 4,
+    frets: 4,
+    chords: ukuleleChords.slice(5).concat(ukuleleChords.slice(0, 5))
+  },
+  guitar: {
+    strings: 6,
+    frets: 5,
+    chords: guitarChords
+  },
+  guitalele: {
+    strings: 6,
+    frets: 5,
+    chords: guitarChords.slice(7).concat(guitarChords.slice(0, 7))
+  }
+};
+// Support for left-handed chord diagrams
+export const reverseString = function(str: string): string {
+  return str.split("").reverse().join("");
+};
+
+type ChordFingeringData = {
+  viewLeft: number;
+  viewWidth: number;
+  width: number;
+  chordName: string;
+  offset?: number;
+  openY: number;
+  dots: { x: number; y: number }[];
+  open: number[];
+  mute: number[];
+} | {
+  chordName: string;
+  unknown: true;
+}
+
+export interface InstrumentData {
+  strings: number;
+  frets: number;
+  chords: { [flavor: string]: string }[];
+}
+
+// Input example:
+// chordName: Bm
+// chordFingering: 4,2,2,2 for G,C,E,A
+// instrumentData: what instrument?
+// Output: SVG rendering of the chord
+export const renderChordFingering = function(chordName: string, chordFingeringStr: string, instrumentData: InstrumentData): ChordFingeringData {
+  const chordFingering = chordFingeringStr.split(",");
+  var offset =
+    chordFingering.every(function(y) {
+      return !(+y > 0) || +y <= instrumentData.frets;
+    }) ?
+      1 :
+      Math.min.apply(null, chordFingering.flatMap(function(y) {
```

Comment:
`import "core-js/fn/array/flat-map"` in any file using `flatMap`, since we need to polyfill it for IE/Edge.